### PR TITLE
Component/pattern

### LINKED ISSUE #24: Pattern
Create the pattern component.

[View on figma](https://www.figma.com/file/gGwSe7ilZCUmAIaJyUc4iY/Bristol-Tech-Fair-Design-System?node-id=289%3A89)

### PR DESCRIPTION
**Pull request by mburridge**
Closes #24

The `<Pattern />` component renders a pattern as specified [here in the design system](https://www.figma.com/file/gGwSe7ilZCUmAIaJyUc4iY/Bristol-Tech-Fair--Design-System?node-id=289%3A89).

I'm unsure how this component needs to interface to the rest of the application, but for this first iteration I've defined two props. One is `color`, this defines the colour of the pattern and takes as its value one of the CSS variables specified in the design system palette. The colours are defined in `Pattern.styles.js` and the `color` prop is added as a class to the first `<path>` element in each SVG.

The other prop is `size`, this defines which of the three pattern widths should be used. Valid values are S, M, and L. Depending on the `size` prop, a `switch statement` selects one of the three SVGs. 

Each of the [nine different patterns in the design system](https://www.figma.com/file/gGwSe7ilZCUmAIaJyUc4iY/Bristol-Tech-Fair--Design-System?node-id=289%3A89) has a different design. So far I have only implement one design for each size, taken from the top row. I'm not sure if I'm on the right track with the approach I've taken so will add the others (if necessary) once this PR has been reviewed.

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -11,6 +12,9 @@ const App = () => {
       <h4>h4</h4>
       <p>p</p>
       <a href="#/">a</a>
+      <Pattern color="midblue" size="M" />{' '}
```
Comment (RobTF9) [inline src/App.jsx]: Remove the component from App. We're using Storybook to view our components so don't need this here.

File: src/components/Pattern/Pattern.stories.jsx [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+import Pattern from './Pattern';
+
+export default {
+  title: 'Components/Pattern',
+  component: Pattern
+};
+
```
Comment (RobTF9) [inline src/components/Pattern/Pattern.stories.jsx]: Add the different variants you want to view in here. For example:

`<Pattern color="midblue" size="M" />`

Will make it easier to preview and test the different variantions.

File: src/components/Pattern/Pattern.jsx [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+      'orange',
+      'lightblue',
+      'midblue',
+      'darkblue'
+    ].includes(color.toLowerCase())
+  ) {
+    patternColor = color.toLowerCase();
+  } else {
+    patternColor = 'midblue';
+  }
```
Comment (RobTF9) [inline src/components/Pattern/Pattern.jsx]: This color if statemant is unnessecary.

You can pass the color prop directly to the styled component.

File: src/components/Pattern/Pattern.jsx [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+          </defs>
+        </svg>
+      `
+      };
+      break;
+    default:
+  }
+
+  return (
+    <PatternWrapper>
```
Comment (RobTF9) [inline src/components/Pattern/Pattern.jsx]: Pass `color` straight in to `PatternWrapper`.

Like so: `<PatternWrapper color={color}>`

File: src/components/Pattern/Pattern.jsx [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+      <div dangerouslySetInnerHTML={svg} />
+    </PatternWrapper>
+  );
+};
+
+Pattern.propTypes = {
+  color: PropTypes.string.isRequired,
```
Comment (RobTF9) [inline src/components/Pattern/Pattern.jsx]: Change color prop type to a oneOf:
```
type: PropTypes.oneOf([      
      'lightpink',
      'darkpink',
      'mattepink',
      'logopink',
      'lightgrey',
      'darkgrey',
      'darkburgundy',
      'orange',
      'lightblue',
      'midblue',
      'darkblue'])
```
You don't need it to be required as you're passing 'midblue' in by default.

File: src/components/Pattern/Pattern.styles.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    fill: var(--orange);
+  }
+  .lightblue {
+    fill: var(--lightblue);
+  }
+  .midblue {
+    fill: var(--midblue);
+  }
+  .darkblue {
+    fill: var(--darkblue);
```
Comment (RobTF9) [inline src/components/Pattern/Pattern.styles.js]: Once you pass `color` in to PatternWrapper you can get rid of all these classes and just add:

```
export const PatternWrapper = styled.div`
   svg {
       fill: ${({color}) => color};
   }   
`;
```

**Review by RobTF9 — DISMISSED** [review]
Hey, good start! Fair amount we can refactor though 😄 

Start by making the color chnages I recommended and then we can look at how the sizing should work 👍 

Reach out on slack if you get stuck.

**Comment by mburridge** [comment]
Hi @RobTF9, I've made the changes requested. A couple of caveats:

1. The SVG doesn't take a background colour, only the first path in the SVG takes a background colour. I've given the first path in each SVG a class and styled that class.
2. Strangely some colours work and others don't, so for example 'darkblue' and 'lightgrey' both work, but 'mattepink' and 'midblue' don't. I can't figure out why some colours work and others don't. 🤷🏻‍♂️

**Comment by shanesc** [comment]
@mburridge Hi Michael, I’ve just been looking through this component to try and learn how to use SVGs - good work so far!
I think the issue with the fill color for certain colors is because you are passing in the string name of the color instead of `var(--mattepink)` for instance.

Try this:
```
export const PatternWrapper = styled.div`
  /* padding: var(--small-space);
  border: 0.1rem solid var(--logopink); */
  .pattern-bg {
    fill: ${({ color }) => `var(--${color})`};
  }
`;
```
Hope that helps, I'm also happy to do some pair programming this weekend if you'd like!

**Comment by mburridge** [comment]
Awesome @shanesc, thanks for that suggestion - that seems to have resolved the issue.

TBH I think I'll struggle to find time this weekend for pair programming as doing some decorating around the home. How about one evening next week?

File: src/components/Pattern/Pattern.jsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    case 'L':
+    case 'l':
+      svg = largeSVG;
+      break;
+    default:
+  }
+
+  return (
+    <PatternWrapper color={color}>
+      <div dangerouslySetInnerHTML={svg} />
```
Comment (shanesc) [inline src/components/Pattern/Pattern.jsx]: I don't have any experience with this. @RobTF9 is the build warning about using this property an issue?

For my own knowledge, why did you use this approach (importing the path as a string of HTML)? I'm so unfamiliar with SVGs so anything I can glean here would be helpful!

**Review by shanesc — APPROVED** [review]
Hi Michael, all looks really good! Since we're just sticking to the three versions for now, this can be merged.

I did add one comment about the `dangerouslySetInnerHTML` property flagging in the Checks, but wasn't sure whether it's something we should be avoiding or not. Once Rob has commented on this I say Squash and Merge!

**Comment by RobTF9** [comment]
@shanesc @mburridge there's no need to use dangerouslySetInnerHTML here.

The best way to refactor this would be to change the Patterns.svg.js file to export three different JSX elements, not strings (we are using React after all)

I can go through this in more detail this evening.

**Comment by mburridge** [comment]
svg's refactored as jsx components to eliminate need for  `dangerouslySetInnerHTML`

File: src/components/Pattern/Pattern.jsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { PatternWrapper } from './Pattern.styles';
+import { SmallSVG, MediumSVG, LargeSVG } from './Pattern.svgs';
+
+const Pattern = ({ color = 'midblue', size = 'M' }) => {
+  // const Pattern = ({ color = 'midblue' }) => {
```
Comment (shanesc) [inline src/components/Pattern/Pattern.jsx]: Delete this comment on line 7

**Review by shanesc — CHANGES_REQUESTED** [review]
Looks great! I like this approach a lot, very clean.

Just delete that one comment line. I think you also have to merge in the most recent commits from `develop` and sort any merge issues, but then it's ready to go 👍

**Comment by admtomas** [comment]
Michael after discusing the Pattern component yesterday we would have to refactor the code.
As it will be just ONE pattern but diferent color then it might be simpler to create Large, Medium, Small components inside Pattern component(please look at Illustration component as reference). This way we just need to pass the color as props, nothing else. 
Pattern component need to inherit `width`  and `height` . For example:  original design of small pattern is nearly twice of the size than we accually need for Resource Card component.
Design of the banners (the funny shapes) is set to grey at the moment but it must inherit background color of the 'component' it would be used. I must stress out that it is imposible to inherit value of grandparent. Rob mentioned about using `fill: transparent` but it will not work neighter. That would leave us with traditional styling by adding `className` to Pattern component and then style it the way we need.

**Comment by mburridge** [comment]
Hi @admtomas 

I'm afraid I don't fully understand what you mean. I'm sorry I had to miss the meeting on Wed so was unable to be party to the discussion that took place. I'll take a look at the Illustration component, as you suggest, and try to figure out the way forward. I'll be sure to ask questions if I need any further clarification.

**Comment by admtomas** [comment]
Michael I did pushed working example to the branch. It is the one i used for testing purposes of Resource Card component. 
I hope it will be helpfull.

File: src/components/Banners/Banners.styles.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  version: '1.1',
+  xmlns: 'http://www.w3.org/2000/svg',
+  xmlnsXlink: 'http://www.w3.org/1999/xlink'
+}))`
+  fill: ${({ color }) => `var(--${color})`};
+  position: absolute;
+  top: 0;
+  right: 0;
+  bottom: 0;
+  left: 0;
```
Comment (shanesc) [inline src/components/Banners/Banners.styles.js]: What's the reason for `position: absolute` here? I was getting issues with it fitting to a parent container, but was resolved once I removed `position: absolute` and the four 0 positions.

Can we remove?

**Review by shanesc — COMMENTED** [review]
Looks great! It's a very clean solution for the different sizes, I like it.

My only comment is about the `position: absolute` on `Banners.styles.js`

File: src/components/Banners/Banners.styles.js (hunk 9 above)
Comment (admtomas) [reply src/components/Banners/Banners.styles.js]: Yes, its gone.

**Review by shanesc — APPROVED** [review]
Looks good, Squash & Merge 👍

**Pull request merged**